lib/vanagon/driver.rb

```diff
@@ -116,7 +118,60 @@ def install_build_dependencies # rubocop:disable Metrics/AbcSize
       end
     end
 
-    def run # rubocop:disable Metrics/AbcSize, Metrics/PerceivedComplexity
+    def build_with_docker # rubocop:disable Metrics/AbcSize
+      if @platform.base_docker_image.nil?
+        raise Vanagon::Error, "The platform must specify the base Docker image that the project will be building from"
+      end
+
+      # Check if the user has Docker installed on their machine.
+      # Vanagon::Docker.docker will raise an error for us if
+      # they don't.
+      #
+      # TODO: We should probably add a has_docker? method to Vanagon::Docker,
+      # but only when we refactor Vanagon::Utilities.find_program_on_path to
+      # _not_ fail (i.e. throw a RuntimeError), but have it raise a Vanagon::Error
+      # instead.
+      Vanagon::Docker.docker("--version")
+
+      remote_workdir = @remote_workdir || "/build"
+      docker_output_dir = "/OUTPUT"
+
+      packaging_workdir = File.join(workdir, "package_build_files")
+      FileUtils.mkdir_p(packaging_workdir)
+
+      @project.fetch_sources(workdir, retry_count, timeout, true)
+      @project.make_bill_of_materials(packaging_workdir)
+      @project.generate_packaging_artifacts(packaging_workdir) unless @project.no_packaging
+      @project.make_docker_build_package_script(packaging_workdir, docker_output_dir)
+      @project.make_dockerfile(workdir, remote_workdir)
+
+      # TODO: We should add the version_from_git helper to the project class
+      # instead. For now, just create a Project::DSL object and call that.
+      project_version = Vanagon::Project::DSL.new(@project.name, @platform).version_from_git
+      docker_name_prefix = "#{@project.name}-#{project_version}"
+      image_tag = "#{docker_name_prefix}:latest"
+
+      Vanagon::Docker.build(workdir, tag: image_tag)
+
+      output_path = File.absolute_path(File.join(VANAGON_ROOT, "output/"))
```

Comment:
Should we use `docker_output_dir` here instead of hard-coding `"output/"`?

Reply:
No -- `docker_output_dir` is the volume mount on the Docker container. This one is the `output/` directory on localhost, which is [hard-coded for the make builds too](https://github.com/puppetlabs/vanagon/blob/master/lib/vanagon/engine/base.rb#L77).

Reply:
yeah, but is there any reason for them to be different?

Reply:
fair.